File: Chess/ChessProblem.cs
namespace Chess
{
	public class ChessProblem
	{
		private static Board board;
		public static ChessStatus ChessStatus;

		public static void LoadFrom(string[] lines)
		{
			board = new Board(lines);
		}

		// Определяет мат, шах или пат белым.
		public static void CalculateChessStatus()
		{
			var isCheck = IsCheckForWhite();
			var hasMoves = false;
			foreach (var locFrom in board.GetPieces(PieceColor.White))
			{
				foreach (var locTo in board.Get(locFrom).Piece.GetMoves(locFrom, board))
				{
					var old = board.Get(locTo);
					board.Set(locTo, board.Get(locFrom));
					board.Set(locFrom, ColoredPiece.Empty);
					if (!IsCheckForWhite())
						hasMoves = true;
					board.Set(locFrom, board.Get(locTo));
					board.Set(locTo, old);
				}
			}
			if (isCheck)
				if (hasMoves)
					ChessStatus = ChessStatus.Check;
				else ChessStatus = ChessStatus.Mate;
			else if (hasMoves) ChessStatus = ChessStatus.Ok;
			else ChessStatus = ChessStatus.Stalemate;
		}

		private static bool IsCheckForWhite()
		{
			var isCheck = false;
			foreach (var loc in board.GetPieces(PieceColor.Black))
			{
				var cell = board.Get(loc);
				var moves = cell.Piece.GetMoves(loc, board);
				foreach (var destination in moves)
				{
					if (board.Get(destination).Is(PieceColor.White, Piece.King))
						isCheck = true;
				}
			}
			if (isCheck) return true;
			return false;
		}
	}
}

File: java/src/main/java/ru/kontur/courses/chess/ChessProblem.java
package ru.kontur.courses.chess;

public class ChessProblem {
    private static Board board;
    public static ChessStatus chessStatus;

    public static void loadFrom(String[] lines) throws Exception {
        board = new BoardParser().parseBoard(lines);
    }

    // Определяет мат, шах или пат белым.
    public static void calculateChessStatus() {
        var isCheck = isCheckForWhite();
        var hasMoves = false;
        for (var locFrom : board.getPieces(PieceColor.WHITE)) {
            for (var locTo : board.getPiece(locFrom).getMoves(locFrom, board)) {
                var old = board.getPiece(locTo);
                board.set(locTo, board.getPiece(locFrom));
                board.set(locFrom, null);
                if (!isCheckForWhite())
                    hasMoves = true;
                board.set(locFrom, board.getPiece(locTo));
                board.set(locTo, old);
            }
        }
        if (isCheck)
            if (hasMoves)
                chessStatus = ChessStatus.CHECK;
            else chessStatus = ChessStatus.MATE;
        else if (hasMoves) chessStatus = ChessStatus.OK;
        else chessStatus = ChessStatus.STALEMATE;
    }

    // check — это шах
    private static boolean isCheckForWhite() {
        var isCheck = false;
        for (var loc : board.getPieces(PieceColor.BLACK)) {
            var piece = board.getPiece(loc);
            var moves = piece.getMoves(loc, board);
            for (var destination : moves) {
                if (Piece.is(board.getPiece(destination), PieceColor.WHITE, PieceType.king))
                    isCheck = true;
            }
        }
        if (isCheck) return true;
        return false;
    }
}
